src/anyp/Uri.cc

```diff
@@ -30,6 +30,45 @@ static const char valid_hostname_chars[] =
     "[:]"
     ;
 
+/// Characters which are valid within a URI userinfo section
+static const CharacterSet &
+UserInfoChars()
+{
+    /*
+     * RFC 3986 section 3.2.1
+     *
+     *  userinfo      = *( unreserved / pct-encoded / sub-delims / ":" )
+     *  unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
+     *  pct-encoded   = "%" HEXDIG HEXDIG
+     *  sub-delims    = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
+     */
+    static const auto userInfoValid = CharacterSet("userinfo", ":-._~%!$&'()*+,;=") +
+                                      CharacterSet::ALPHA +
+                                      CharacterSet::DIGIT;
+    return userInfoValid;
+}
+
+/**
+ * Governed by RFC 3986 section 2.1
+ */
+SBuf
+AnyP::Uri::Encoder(const SBuf &buf, const CharacterSet &encode)
+{
+    if (buf.isEmpty())
+        return buf;
+
+    SBuf output;
+    output.reserveSpace(buf.length()*3); // worst-case every byte is encoded
```

Comment:
The only optimization that I ma aware of to enable returning buf, is a full-length scan of the string when no characters need encoding. Using FindFirstNotOf() will prohibit using range-for the jumps are not equal in length.

Reply:
Whether this code is using a for-range loop is unimportant. There are many ways to optimize this. Here are two candidate sketches:

1. Simpler with a bit of code duplication:
```C++
SBuf goodSection;
// optimization for the arguably common "no encoding necessary" case
if (tk.prefix(goodSection, goodChars) && tk.atEnd())
    return input;
// else fall through the reserveSpace() to the append() below it
output.reserveSpace(input.length() * 3); // worst case: encode all chars
output.append(goodSection); // may be empty

while (!tk.atEnd()) {
    // TODO: Add Tokenizer::parseOne(void).
    const auto ch = tk.remaining()[0]; 
    output.appendf("%%%02X", static_cast<unsinged int>(ch)); // TODO: Optimize using a table
    (void)tk.skipOne(ch);

    if (tk.prefix(goodSection, goodChars))
        output.append(goodSection);
}
```

2. A bit more complex but without code duplication:
```C++
const auto reserveIfNeeded = [&input, &output]() -> void {
    if (!output.length())
        output.reserveSpace(input.length() * 3); // worst case: encode all chars
}
while (!tk.atEnd()) {
    SBuf goodSection;
    if (tk.prefix(goodSection, goodChars)) {
        // optimization for the arguably common "no encoding necessary" case
        if (goodSection.length() == input.length())
            return input; 
        reserveIfNeeded();
        output.append(goodSection);
    }
    if (!tk.atEnd()) {
        // TODO: Add Tokenizer::parseOne(void).
        const auto ch = tk.remaining()[0]; 
        reserveIfNeeded();
        output.appendf("%%%02X", static_cast<unsinged int>(ch)); // TODO: Optimize using a table
        (void)tk.skipOne(ch);
    }
}
```

Implementing the documented `Tokenizer::parseOne()` in this PR is fine but is not required IMO. The new method should assert on empty input. It will make the encoding `if` statement a lot simpler.

The above asymmetrical sketches assume the probability of a character encoding is less than 50%. It is possible to write a symmetrical implementation that would assume that the probability of a character encoding is about 50%; that implementation would not need a Tokenizer::parseOne() call but will need two character sets as a parameter unless we enhance our CharacterSet to prebuild and store the complement set internally. It is also possible to write an asymmetrical implementation that assumes that the probability of a character encoding is above 50%.

Reply:
I have added a unit benchmark in commit 886eaf3bb4326b6a678b05804bb7e587a33ee768 for the function comparing the simple for-loop scan vs whatever our current Encode implementation is. The benchmark shows both of your implementations are within CPU load variance of each other - so going with least complex  in commit f422f38c907bfd2fb20bc9bde261447771dfb728

Best/Worst results from a few runs of each algorithm.

Algorithm 1:

AnyP::Uri::Encode non-change: baseline= 0.339505 , encoder= 0.017508
AnyP::Uri::Encode all changed: baseline= 108.925 , encoder= 3.83041

AnyP::Uri::Encode non-change: baseline= 0.314258 , encoder= 0.025682
AnyP::Uri::Encode all changed: baseline= 99.7914 , encoder= 4.12767


Algorithm 2:

AnyP::Uri::Encode non-change: baseline= 0.312348 , encoder= 0.017203
AnyP::Uri::Encode all changed: baseline= 100.431 , encoder= 3.81312

AnyP::Uri::Encode non-change: baseline= 0.336105 , encoder= 0.017345
AnyP::Uri::Encode all changed: baseline= 100.508 , encoder= 4.35364